Have models inherit from base class, support model polymorphism (continuation of  #128)

macisamuele opened the pull request:
This CR enrich the great modification provided by @jlumpe in #128.
* I have merged master branch solved conflicts and re-added ``include_missing_properties`` configuration
* Added tests for discriminator feature.

sjaensch commented on CHANGELOG.rst:
```diff
@@ -5,6 +5,8 @@ X.X.X (XXXX-XX-XX)
 ------------------
 - Add include_missing_properties configuration - PR #152
 - Consider default when unmarshalling - PR #154
+- Refactor of models generation - PR #128, #159. Thanks Michael Jared Lumpe for your contribution
```
Not sure this requires its own entry in the changelog. Would the next line be enough? Maybe add it in there if you want to mention it?

macisamuele replied:
Removed it!

sjaensch commented on bravado_core/schema.py:
```diff
@@ -66,6 +66,56 @@ def is_list_like(spec):
+        # Found a single type
+        return found_types.pop()
+
+    if found_types == {'integer', 'number'}:
```
I'm not sure this is correct. Doesn't this mean what we have here is an object composed of two properties, both of which are integers? I.e. in almost every "allOf" case the type should be "object", right?

sjaensch replied:
We should also add a test for the allOf case here.

macisamuele replied:
You're completely right ... as far as I'm understanding in case of ``allOf`` the type should be set to ``object`` (according to [Swagger Specs > Composition and Inheritance (Polymorphism)](http://swagger.io/specification/)).

I'm modifying it and adding the needed tests

sjaensch commented on bravado_core/swagger20_validator.py:
```diff
@@ -104,6 +106,59 @@ def enum_validator(swagger_spec, validator, enums, instance, schema):
+    new_schema = deepcopy(swagger_spec.deref(swagger_spec.spec_dict['definitions'][discriminator_value]))
+    if 'allOf' not in new_schema:
+        raise ValidationError(
+            message='discriminated schema \'{}\' must inherits from \'{}\''.format(
```
must inherit from

sjaensch commented on bravado_core/swagger20_validator.py:
```diff
@@ -104,6 +106,59 @@ def enum_validator(swagger_spec, validator, enums, instance, schema):
+    if not schemas_to_remove:
+        # Not checking against len(schemas_to_remove) > 1 because it should be prevented by swagger spec validation
+        raise ValidationError(
+            message='discriminated schema \'{}\' must inherits from \'{}\''.format(
```
same here

sjaensch commented on bravado_core/swagger20_validator.py:
```diff
@@ -104,6 +106,59 @@ def enum_validator(swagger_spec, validator, enums, instance, schema):
+        )
+
+    new_schema = deepcopy(swagger_spec.deref(swagger_spec.spec_dict['definitions'][discriminator_value]))
+    if 'allOf' not in new_schema:
```
Are you sure this is always correct? What if you have an instance of the base object?

macisamuele replied:
You're right ... I'm adding a test for this case

macisamuele replied:
done

sjaensch commented on bravado_core/swagger20_validator.py:
```diff
@@ -104,6 +106,59 @@ def enum_validator(swagger_spec, validator, enums, instance, schema):
         yield error
 
 
+def discriminator_validator(swagger_spec, validator, discriminator_attribute, instance, schema):
```
This method needs unit tests. :)

macisamuele replied:
This is already tested by [``test_validate_valid_polymorphic_object``](https://github.com/macisamuele/bravado-core/blob/models/tests/validate/validate_object_test.py#L339) and [``test_validate_invalid_polymorphic_object``](https://github.com/macisamuele/bravado-core/blob/models/tests/validate/validate_object_test.py#L401)

sjaensch commented on bravado_core/model.py:
```diff
@@ -69,142 +67,321 @@ def collect_models(container, key, path, models, swagger_spec):
+    def _as_dict(self, additional=False, recursive=False):
+        """Get property values as dictionary.
+
+        :param bool additional: Whether to include additional properties
```
I'd name this either additional_props (see above) or additional_properties.

sjaensch commented on bravado_core/model.py:
```diff
@@ -69,142 +67,321 @@ def collect_models(container, key, path, models, swagger_spec):
+        """Names of properties in instance which are not defined in spec."""
+        return set(self.__dict).difference(self._properties)
+
+    def _as_dict(self, additional=False, recursive=False):
```
I would assume that `recursive=True` is the expected behavior, can we change that?

macisamuele replied:
sure

sjaensch commented on tests/model/model_test.py:
```diff
@@ -0,0 +1,84 @@
+
+    assert 'id' in set(user)
+    del user.id
+    assert 'id' in set(user)
```
Can you explain a bit what we're testing here? It's not easy to understand if you don't read `model.py` at the same time, especially since you don't assert on the user.id value before and after. I.e. maybe say "deleting a property defined in the spec should set the value to `None`, and assert on the value before and after.

macisamuele replied:
👍  adding the comment

sjaensch commented on tests/model/model_test.py:
```diff
@@ -0,0 +1,84 @@
+        user[property] == user_kwargs.get(property)
+        for property in definitions_spec['User']['properties'].keys()
+        if property != 'id'
+    )
```
What is this asserting exactly? The any() makes it really muddy. It says that at least one of the user_kwargs values needs to be equal to the one in the user dict...?

macisamuele replied:
wrong function ... it should be ``all``.
I'm checking that each property (except ``id``) has the same value either in the model (``user``) and in the input dictionary (``user_kwargs``)

sjaensch commented on tests/model/model_test.py:
```diff
@@ -0,0 +1,84 @@
+    user_type._model_spec['additionalProperties'] = True
+    user = user_type(foo='bar', **user_kwargs)
+
+    assert 'foo' in set(user)
```
I don't think you need the set()...? If you do then maybe it's better to just use user.keys()?

macisamuele replied:
removing ``set`` and ``.keys()`` since not needed

macisamuele commented:
* Changed ``additional_properties`` and ``recursive`` default values
* Added tests for ``_as_dict`` with and without recursion
* Updated ``_as_dict`` in order to go recursively through lists

sjaensch commented on tests/model/model_test.py:
```diff
@@ -70,6 +74,43 @@ def test_model_as_dict(definitions_spec, user_type, user_kwargs):
+        False,
+    ]
+)
+def test_marsshal_as_dict_recursive(polymorphic_spec, recursive):
```
marshal
